### PR TITLE
[Test/Sample test] Fix the imparity between experiment name setting of notebook samples and python samples

### LINKED ISSUE #2148: Sample test infra: inconsistent behavior of dynamically injected arguments for output and experiment_name
The current behavior is:
The output argument is injected for both normal python sample and notebook sample when it is specified in the config yaml. 
The experiment_name is only injected for notebook samples and will be injected even though it is not specified in the config yaml.

Expected behavior:
Both are injected only when they are specified in the config yaml. Since the assumption is that the notebook submits a job programmatically while the python samples do not, experiment_name arg will not be for notebook samples.

### PR DESCRIPTION
**Pull request by numerology**
Fix #2148

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2362)

### CODE REVIEW

**Comment by Ark-kun** [comment]
Maybe it would be easier to just fix https://github.com/kubeflow/pipelines/pull/2292
I'll need a bit more time to debug it.

File: test/sample-test/sample_test_launcher.py [hunk 1]
```diff
@@ -129,7 +129,9 @@ def _compile(self):
           self._run_pipeline = raw_args['run_pipeline']
 
       if self._run_pipeline:
-        nb_params['experiment_name'] = self._test_name + '-test'
+        if not raw_args['notebook_params']['experiment_name']:
```
Comment (gaoning777) [inline test/sample-test/sample_test_launcher.py]: Could you move all the yaml validation codes in one function, including the yamale validate?

File: test/sample-test/sample_test_launcher.py [hunk 2]
```diff
@@ -129,7 +129,9 @@ def _compile(self):
           self._run_pipeline = raw_args['run_pipeline']
```
Comment (gaoning777) [inline test/sample-test/sample_test_launcher.py]: Could you add some comments in the schema yaml about the run_pipeline:
1) for python samples: setting it true means submitting the pipeline.
2) for notebook samples: setting it true means telling the test infra that the sample will submit a pipeline run and the test infra will inject the experiment_name.

Comment (gaoning777) [reply test/sample-test/sample_test_launcher.py]: In other words: run_pipeline means submitting the run for python sample and checking the results in the notebook sample.

File: test/sample-test/sample_test_launcher.py [hunk 3]
```diff
@@ -129,7 +129,9 @@ def _compile(self):
...
 
       if self._run_pipeline:
-        nb_params['experiment_name'] = self._test_name + '-test'
+        if not raw_args['notebook_params']['experiment_name']:
+          raise ValueError('Experiment name is required if pipeline run is expected.')
+        nb_params['experiment_name'] = raw_args['notebook_params']['experiment_name']
```
Comment (gaoning777) [inline test/sample-test/sample_test_launcher.py]: The fact that the experiment_name value in the config yaml is not used(but generated by the test infra) should be noted in the schema yaml file, similar to the output value not being used. WDYT?

**Comment by gaoning777** [comment]
> /hold further work needed.

Could you specify what further work is required? Maybe put a link here?

**Comment by numerology** [comment]
> > /hold further work needed.
> 
> Could you specify what further work is required? Maybe put a link here?

Sure. Since #2292 is done, I think it will be beneficial to incorporate the newly introduced mechanism here, which will reduce the need of injecting experiment names.

**Comment by gaoning777** [comment]
> > > /hold further work needed.
> > 
> > 
> > Could you specify what further work is required? Maybe put a link here?
> 
> Sure. Since #2292 is done, I think it will be beneficial to incorporate the newly introduced mechanism here, which will reduce the need of injecting experiment names.

I think it is beneficial to explicitly add the experiment_name in the notebook examples, though. Or else, what is the default experiment_name when the env_var is not configured? WDYT?

**Comment by numerology** [comment]
> > > /hold further work needed.
> > 
> > 
> > Could you specify what further work is required? Maybe put a link here?
> 
> Sure. Since #2292 is done, I think it will be beneficial to incorporate the newly introduced mechanism here, which will reduce the need of injecting experiment names.

Per my reading it should be 'Default'. Sure, we can specify the experiment name in the notebook, but IMO changing it through env var will allow us to get rid of some ad hoc logic in current sample test infra. For example, we do not need to decide whether to inject experiment_name or not based on run_pipeline flag, instead we just set the env var when executing the pipeline.

**Comment by gaoning777** [comment]
> > > > /hold further work needed.
> > > 
> > > 
> > > Could you specify what further work is required? Maybe put a link here?
> > 
> > 
> > Sure. Since #2292 is done, I think it will be beneficial to incorporate the newly introduced mechanism here, which will reduce the need of injecting experiment names.
> 
> Per my reading it should be 'Default'. Sure, we can specify the experiment name in the notebook, but IMO changing it through env var will allow us to get rid of some ad hoc logic in current sample test infra. For example, we do not need to decide whether to inject experiment_name or not based on run_pipeline flag, instead we just set the env var when executing the pipeline.

I'm all up for supporting the env-var. My point is for the experiment_name to be explicit in the samples. In terms of the ad hoc logic, isn't injecting the environment variable no different from adding the parameter to the papermill?

**Comment by numerology** [comment]
> > > > > /hold further work needed.
> > > > 
> > > > 
> > > > Could you specify what further work is required? Maybe put a link here?
> > > 
> > > 
> > > Sure. Since #2292 is done, I think it will be beneficial to incorporate the newly introduced mechanism here, which will reduce the need of injecting experiment names.
> > 
> > 
> > Per my reading it should be 'Default'. Sure, we can specify the experiment name in the notebook, but IMO changing it through env var will allow us to get rid of some ad hoc logic in current sample test infra. For example, we do not need to decide whether to inject experiment_name or not based on run_pipeline flag, instead we just set the env var when executing the pipeline.
> 
> I'm all up for supporting the env-var. My point is for the experiment_name to be explicit in the samples. In terms of the ad hoc logic, isn't injecting the environment variable no different from adding the parameter to the papermill?

One difference, for example, is that we can set the env var no matter what, but we can only inject experiment_name to notebook when it's indeed needed.

**Comment by gaoning777** [comment]
I don't think papermill would fail if there is no such parameters. What it does is to inject a cell below the cell with the 'parameters' tag. Therefore, we can inject the experiment_name no matter what.

**Comment by numerology** [comment]
> I don't think papermill would fail if there is no such parameters. What it does is to inject a cell below the cell with the 'parameters' tag. Therefore, we can inject the experiment_name no matter what.

You're right... Somewhat I got the wrong impression that it would raise errors. Now that it does not, I think I can further simplify the code a little bit.

**Comment by Ark-kun** [comment]
> I think it is beneficial to explicitly add the experiment_name in the notebook examples, though.

Let's brainstorm on the pros and cons of having `experiment_name` in the notebooks.
Some of the cons that come to mind:
* Extra concept for users to learn (the "Experiment" concept).
* If the experiment names are different then we're going to "spam" the "Experiments" section of the user's cluster.
* Currently there is no way to delete experiments in the UX.
* Extra line/configuration, which is not required for running the sample. The samples should be as and focused small as possible.

**Comment by gaoning777** [comment]
> > I think it is beneficial to explicitly add the experiment_name in the notebook examples, though.
> 
> Let's brainstorm on the pros and cons of having `experiment_name` in the notebooks.
> Some of the cons that come to mind:
> 
> * Extra concept for users to learn (the "Experiment" concept).
> * If the experiment names are different then we're going to "spam" the "Experiments" section of the user's cluster.
> * Currently there is no way to delete experiments in the UX.
> * Extra line/configuration, which is not required for running the sample. The samples should be as and focused small as possible.

I think it is fine to remove the experiment_name and use the default experiment in the samples. And, https://github.com/kubeflow/pipelines/pull/2292 is also merged.

**Comment by gaoning777** [comment]
@numerology, could you resolve the comments and we can merge this PR?

**Comment by numerology** [comment]
> @numerology, could you resolve the comments and we can merge this PR?

After merging #2292 it seems like there is not too much to do in this PR. WDYT?
After #2292 it will be the case that for both python and notebook samples the experiment names are set via env var no matter what.

**Comment by gaoning777** [comment]
SGTM

**Comment by gaoning777** [comment]
Feel free to close this.

**Comment by numerology** [comment]
/close
Per discussion

**Pull request closed without merging**